### PR TITLE
[bitnami/postgresql-ha] Don't render empty keys from templates

### PR DESCRIPTION
**Pull request by tgh19**
Signed-off-by: Trey Hoffman treyhoffman@outlook.com

### Description of the change

This change prevents rendered manifests from containing empty keys for `volumes` and `initContainers`. If there are no volumes or initContainers, then the rendered manifests should exclude their keys.

### Benefits

Datree tests would fire false positives if these empty keys are found in the manifests. Plus, it does not make sense to include a key for initContainers if there are no initContainers that actually need to run.

### Possible drawbacks

Creates somewhat more complex templating code.

### Applicable issues

NA

### Additional information

NA

### Checklist

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [X] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [X] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

**Comment by tgh19** [comment]
@javsalgar , in The previous PR (https://github.com/bitnami/charts/pull/16575) I had incorrect commit signing. I closed that one however, this PR did not automatically assign a reviewer. Could you take a look whenever you have time, please?

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml [hunk 1]
```diff
@@ -98,10 +100,10 @@ spec:
               mountPath: /tmp/certs
             - name: pgpool-certificates
               mountPath: /opt/bitnami/pgpool/certs
-      {{- end }}
       {{- if .Values.pgpool.initContainers }}
       {{- include "common.tplvalues.render" (dict "value" .Values.pgpool.initContainers "context" $) | nindent 8 }}
       {{- end }}
+      {{- end }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/pgpool/deployment.yaml:106]: Moving this `{{- end}}` would nest the `pgpool.initContainers` under the `pgpool.tls.enabled`, causing the first to not be rendered if the later is not enabled.

```
      {{- if .Values.pgpool.tls.enabled }}
        - name: init-chmod-data
          image: {{ template "postgresql-ha.volumePermissions.image" . }}
              mountPath: /tmp/certs
            - name: pgpool-certificates
              mountPath: /opt/bitnami/pgpool/certs
      {{- if .Values.pgpool.initContainers }}
      {{- include "common.tplvalues.render" (dict "value" .Values.pgpool.initContainers "context" $) | nindent 8 }}
      {{- end }}
      {{- end }}
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Hi @tgh19,

Thank you very much for your contribution!

I see several issues in this PR related to the `if/end` closing. Could you please refactor it to follow this style?
```
{{- if or .Values.feature1 .Values.feature2 ... }}
initContainers:
{{- if .Values.feature1 }}
   - name: feature1
{{- end }} #Close feature1
{{- if .Values.feature2 }}
   - name: feature2
{{- end }} #Close feature2
...
{{- end }} #Close OR conditional
```
This is the style used in other charts in the catalog, and I'm worried the proposed style could cause other issues, similar to what happened with the `init-chmod-data` init container.

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml (hunk 1 above)
Comment (tgh19) [reply bitnami/postgresql-ha/templates/pgpool/deployment.yaml:106]: Thank you for the feedback. This should be resolved now. Please let me know if you find any additional issues!

File: bitnami/postgresql-ha/templates/postgresql/statefulset.yaml [hunk 2]
```diff
@@ -80,6 +80,11 @@ spec:
       {{- end }}
       hostNetwork: {{ .Values.postgresql.hostNetwork }}
       hostIPC: {{ .Values.postgresql.hostIPC }}
+      {{- if .Values.postgresql.tls.enabled }}
+      initContainers:
+      {{- else if and .Values.volumePermissions.enabled (or (or (not (empty .Values.postgresql.extendedConf)) (not (empty .Values.postgresql.extendedConfCM))) .Values.persistence.enabled) }}
+      initContainers:
+      {{- else if or .Values.postgresql.initContainers .Values.postgresql.extraInitContainers }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/postgresql/statefulset.yaml:87]: Could you please reduce this to a single conditional?

```suggestion
{{- if or .Values.postgresql.tls.enabled .Values.postgresql.initContainers .Values.postgresql.extraInitContainers (and .Values.volumePermissions.enabled (or (or (not (empty .Values.postgresql.extendedConf)) (not (empty .Values.postgresql.extendedConfCM))) .Values.persistence.enabled)) }}
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
I left some comments, could you please take a look at them?

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml [hunk 3]
```diff
@@ -331,7 +333,9 @@ spec:
           {{- if .Values.pgpool.resources }}
           resources: {{- toYaml .Values.pgpool.resources | nindent 12 }}
           {{- end }}
+          {{- if or .Values.pgpool.configuration .Values.pgpool.configurationCM .Values.pgpool.initdbScripts .Values.pgpool.initdbScriptsCM .Values.pgpool.initdbScriptsSecret .Values.postgresql.usePasswordFile .Values.pgpool.usePasswordFile .Values.pgpool.tls.enabled .Values.pgpool.extraVolumeMounts }}
           volumeMounts:
+          {{- end }}
```
[the remark this answers is not in the document]
Comment (tgh19) [reply bitnami/postgresql-ha/templates/pgpool/deployment.yaml]: Absolutely. Let me know if there's anything else. Thanks for your patience.

File: bitnami/postgresql-ha/templates/postgresql/statefulset.yaml [hunk 4]
```diff
@@ -80,6 +80,7 @@ spec:
       {{- end }}
       hostNetwork: {{ .Values.postgresql.hostNetwork }}
       hostIPC: {{ .Values.postgresql.hostIPC }}
+{{- if or .Values.postgresql.tls.enabled .Values.postgresql.initContainers .Values.postgresql.extraInitContainers (and .Values.volumePermissions.enabled (or (or (not (empty .Values.postgresql.extendedConf)) (not (empty .Values.postgresql.extendedConfCM))) .Values.persistence.enabled)) }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/postgresql/statefulset.yaml:83]: I'm sorry for being picky with this one, could you please indent this line?

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Hi there, the changes look good to me, but could you please address one last suggestion?

File: bitnami/postgresql-ha/templates/postgresql/statefulset.yaml (hunk 4 above)
Comment (tgh19) [reply bitnami/postgresql-ha/templates/postgresql/statefulset.yaml:83]: Updated!

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml [hunk 5]
```diff
@@ -364,9 +367,11 @@ spec:
             {{- if .Values.pgpool.extraVolumeMounts }}
             {{- include "common.tplvalues.render" (dict "value" .Values.pgpool.extraVolumeMounts "context" $) | nindent 12 }}
             {{- end }}
+          {{- end }}
         {{- if .Values.pgpool.sidecars }}
         {{- include "common.tplvalues.render" (dict "value" .Values.pgpool.sidecars "context" $) | nindent 8 }}
         {{- end }}
+      {{- if or .Values.pgpool.configuration .Values.pgpool.configurationCM .Values.pgpool.initdbScripts .Values.pgpool.initdbScriptsCM .Values.pgpool.initdbScriptsSecret .Values.postgresql.usePasswordFile .Values.pgpool.tls.enabled .Values.pgpool.extraVolumes }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/pgpool/deployment.yaml:374]: This conditional is missing the value `.Values.pgpool.usePasswordFile`.

**Review by migruiz4 — CHANGES_REQUESTED** [review]
I have found a small issue in one of the conditionals, could you please address it?

**Comment by bfritz** [comment]
Looks like there might be another case of this when `.Values.metrics.enabled` evaluates true but both `.Values.postgresql.usePasswordFile` and `.Values.metrics.customMetrics` evaluate false:
https://github.com/bitnami/charts/pull/16576/files#diff-43f8bd5eb07892aa159485018e8bec5d30fc810e8713c5f421111ca8cccfa877R525-R533

**Comment by migruiz4** [comment]
Nice catch @bfritz!

@tgh19 when you have the time to continue working on this PR, could you please also apply that suggestion?

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml (hunk 5 above)
Comment (tgh19) [reply bitnami/postgresql-ha/templates/pgpool/deployment.yaml]: Updated and I also included some multi-line formatting that makes this more apparent..

**Comment by tgh19** [comment]
@migruiz4 , when you have a moment, could you take another look at this?

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml [hunk 6]
```diff
@@ -331,6 +333,15 @@ spec:
...
           {{- end }}
+          {{- if or
+            (or .Values.pgpool.configuration .Values.pgpool.configurationCM)
+            (or .Values.pgpool.initdbScripts .Values.pgpool.initdbScriptsCM)
+            (.Values.pgpool.initdbScriptsSecret)
+            (.Values.postgresql.usePasswordFile)
+            (.Values.pgpool.usePasswordFile)
+            (.Values.pgpool.tls.enabled)
+            (.Values.pgpool.extraVolumeMounts)
+          }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/pgpool/deployment.yaml:344]: Could you please not split this into multiple lines?
I know that a single super-long line isn't pretty or precisely easy to read, but at the moment I prefer sticking with the style used in the rest of the catalog.
In the future, we may implement some linting guidelines and extend it to all the charts in the catalog, including here.

File: bitnami/postgresql-ha/templates/pgpool/deployment.yaml [hunk 7]
```diff
@@ -364,9 +375,19 @@ spec:
...
         {{- end }}
+      {{- if or
+        (or .Values.pgpool.configuration .Values.pgpool.configurationCM)
+        (or .Values.pgpool.initdbScripts .Values.pgpool.initdbScriptsCM)
+        (.Values.pgpool.initdbScriptsSecret)
+        (.Values.pgpool.usePasswordFile)
+        (.Values.postgresql.usePasswordFile)
+        (.Values.pgpool.tls.enabled)
+        (.Values.pgpool.extraVolumes)
+      }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/pgpool/deployment.yaml:390]: Same comment on multi-lines.

File: bitnami/postgresql-ha/templates/postgresql/statefulset.yaml [hunk 8]
```diff
@@ -80,6 +80,12 @@ spec:
       {{- end }}
       hostNetwork: {{ .Values.postgresql.hostNetwork }}
       hostIPC: {{ .Values.postgresql.hostIPC }}
+      {{- if or
+        (.Values.postgresql.tls.enabled)
+        (.Values.postgresql.initContainers)
+        (.Values.postgresql.extraInitContainers)
+        (and .Values.volumePermissions.enabled (or (or (not (empty .Values.postgresql.extendedConf)) (not (empty .Values.postgresql.extendedConfCM))) .Values.persistence.enabled))
+      }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/postgresql/statefulset.yaml:88]: Same comment on multi-lines.

**Comment by tgh19** [comment]
@migruiz4 , this is ready to go whenever you have a moment. Thanks again for your patience.

**Comment by tgh19** [comment]
@migruiz4 , are you around to take a look at this one please?

File: bitnami/postgresql-ha/templates/postgresql/statefulset.yaml [hunk 9]
```diff
@@ -533,6 +536,7 @@ spec:
             - name: custom-metrics
               mountPath: /conf
               readOnly: true
+          {{- end }}
```
Comment (migruiz4) [inline bitnami/postgresql-ha/templates/postgresql/statefulset.yaml:539]: There is an issue with the if/end here, could you please apply this suggestion?
```suggestion
            {{- end }}
          {{- end }}
          {{- if .Values.metrics.customMetrics }}
```

This is to separate the conditional from `volumeMounts` and `command`/`args` when `metrics.customMetrics` are enabled.

This way, all if blocks should be properly closed:

```
          {{- if or .Values.postgresql.usePasswordFile .Values.metrics.customMetrics }} #1
          volumeMounts:
            {{- if .Values.postgresql.usePasswordFile }} #2
            - name: password
              mountPath: /opt/bitnami/postgresql/secrets/
            {{- end }} #2
            {{- if .Values.metrics.customMetrics }} #3
            - name: custom-metrics
              mountPath: /conf
              readOnly: true
            {{- end }} #3
          {{- end }} #1
          {{- if .Values.metrics.customMetrics }} #4
          {{- if .Values.diagnosticMode.enabled }} #5
          command: {{- include "common.tplvalues.render" (dict "value" .Values.diagnosticMode.command "context" $) | nindent 12 }}
          {{- end }} #5
          {{- if .Values.diagnosticMode.enabled }} #6
          args: {{- include "common.tplvalues.render" (dict "value" .Values.diagnosticMode.args "context" $) | nindent 12 }}
          {{- else }} #6
          args: ["--extend.query-path", "/conf/custom-metrics.yaml"]
          {{- end }} #6
          {{- end }} #4
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Please apply the last suggestion and we should be ready to merge!

**Review by migruiz4 — APPROVED** [review]
Looks good to me! Thank you for your contribution!

**Pull request merged**